lib/Sema/CodeSynthesis.cpp

```diff
@@ -1183,6 +1184,235 @@ static FuncDecl *completeLazyPropertyGetter(VarDecl *VD, VarDecl *Storage,
+  
+  auto InitStorageRef = new (Context) DeclRefExpr(SpecializeInitStorage,
+                                                  DeclNameLoc(),
+                                                  /*implicit*/ true);
+  auto InitStorageMethodTy = FunctionType::get(Context.TheEmptyTupleType,
+                                               SubstStorageContextTy);
+  auto InitStorageRefTy = FunctionType::get(SelfTypeRef->getType(),
+                                            InitStorageMethodTy);
+  InitStorageRef->setType(InitStorageRefTy);
+
+  auto SelfApply = new (Context) DotSyntaxCallExpr(InitStorageRef,
+                                                   SourceLoc(),
+                                                   SelfTypeRef);
+  SelfApply->setImplicit();
+  SelfApply->setType(InitStorageMethodTy);
+  auto EmptyTuple = TupleExpr::create(Context,
+                                      SourceLoc(), {},{},{}, SourceLoc(),
+                                      /*trailing closure*/ false,
+                                      /*implicit*/ true,
+                                      Context.TheEmptyTupleType);
+  auto InitStorageExpr = new (Context) CallExpr(SelfApply, EmptyTuple,
+                                                /*implicit*/ true);
+  InitStorageExpr->setImplicit();
+  InitStorageExpr->setType(SubstStorageContextTy);
+  
+  // Create the pattern binding decl for the storage decl.  This will get
+  // default initialized using the protocol's initStorage() method.
+  Pattern *PBDPattern = new (Context) NamedPattern(Storage, /*implicit*/true);
+  PBDPattern = new (Context) TypedPattern(PBDPattern,
+                                          TypeLoc::withoutLoc(SubstStorageInterfaceTy),
+                                          /*implicit*/true);
+  auto *PBD = PatternBindingDecl::create(Context, /*staticloc*/SourceLoc(),
+                             VD->getParentPatternBinding()->getStaticSpelling(),
+                             /*varloc*/VD->getLoc(),
+                             PBDPattern, /*init*/InitStorageExpr,
+                             VD->getDeclContext());
+  PBD->setImplicit();
+  PBD->setInitializerChecked(0);
+  addMemberToContextIfNeeded(PBD, VD->getDeclContext(), VD);
+  
+  // Mark the vardecl to be final, implicit, and private.  In a class, this
+  // prevents it from being dynamically dispatched.
+  if (VD->getDeclContext()->getAsClassOrClassExtensionContext())
+    makeFinal(Context, Storage);
+  Storage->setImplicit();
+  Storage->setAccessibility(Accessibility::Private);
+  Storage->setSetterAccessibility(Accessibility::Private);
+  
+  // Add accessors to the storage, since we'll need them to satisfy the
+  // conformance requirements.
+  addTrivialAccessorsToStorage(Storage, *this);
+  
+  // Add the witnesses to the conformance.
+  ArrayRef<Substitution> MemberSubs;
+  if (DC->isTypeContext() && DC->isGenericContext()) {
+    MemberSubs = DC->getDeclaredTypeInContext()->getAs<BoundGenericType>()
```

Comment:
Is DC->getGenericParamsOfContext()->getForwardingSubstitutions() more correct if we ever do nested generics?